### PR TITLE
Add arbitrary folder validation

### LINKED ISSUE #547: Validate arbitrary path before running the checks
Follow-up of [Allow static checks to be run against an arbitrary path ](https://github.com/WordPress/plugin-check/pull/490)

Proper validation should be done before start checking arbitrary path. If not a proper WordPress plugin folder, checks should not run.

### PR DESCRIPTION
**Pull request by ernilambar**
Fixes https://github.com/WordPress/plugin-check/issues/547

### CODE REVIEW

File: includes/Checker/Abstract_Check_Runner.php [hunk 1]
```diff
@@ -477,16 +477,50 @@ private function get_check_exclude_slugs() {
...
 	 */
 	final public function get_plugin_basename() {
 		if ( null === $this->plugin_basename ) {
 			$plugin = null !== $this->plugin ? $this->plugin : $this->get_plugin_param();
 
-			if ( is_dir( $plugin ) ) {
-				$this->plugin_basename = $plugin;
-			} else {
+			// Check if a slug.
+			if ( preg_match( '/^[a-z0-9]+(-?[a-z0-9]+)*(\.php)?$/i', $plugin ) ) {
```
Comment (swissspidy) [inline includes/Checker/Abstract_Check_Runner.php]: A regex check is error prone. Can you elaborate on this?

Why not also just check if it contains a plugin header?

File: includes/Checker/Abstract_Check_Runner.php [hunk 2]
```diff
@@ -477,16 +477,50 @@ private function get_check_exclude_slugs() {
...
+						if ( ! empty( $plugin_data['Name'] ) ) {
+							$main_file_found = true;
+							break;
+						}
+					}
+
+					if ( true === $main_file_found ) {
+						$this->plugin_basename = $plugin;
+					} else {
+						throw new Exception(
```
Comment (swissspidy) [inline includes/Checker/Abstract_Check_Runner.php]: Is this method the right place to do this kind of validation?

Why not move this to `Plugin_Request_Utility`?

File: includes/Checker/Abstract_Check_Runner.php (hunk 1 above)
Comment (ernilambar) [reply includes/Checker/Abstract_Check_Runner.php]: We could have inputs like this:
```
sample-slug
hello.php
/var/www/path/plugin/
https://example.com/files/plugin-slug.zip (In the future PR)
```
This is for checking which routes should we use for finding basename.

Comment (ernilambar) [reply includes/Checker/Abstract_Check_Runner.php]: Now wondering how WP-CLI handles this. 😊

Comment (ernilambar) [reply includes/Checker/Abstract_Check_Runner.php]: CLI does not seem to accept folder as input. `wp plugin install` **plugin|zip|url** - Accepts a plugin slug, the path to a local zip file, or a URL to a remote zip file.

Comment (ernilambar) [reply includes/Checker/Abstract_Check_Runner.php]: Removed the regex.

**Review by swissspidy — APPROVED** [review]

**Review by davidperezgar — APPROVED** [review]

**Pull request merged**